rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,329 @@
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// the Clock to use for insertion timestamps
+    ///
+    /// default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
+
+    /// the default time-to-live (TTL) for insertions (note that re-inserting a key
+    /// with any value will update the TTL to use the default_ttl)
+    ///
+    /// default: no TTL (insertions will remain until deleted)
+    default_ttl: Option<Duration>
+}
+```
+
+To enforce that the clock remains monotonically increasing across instantiations of the
+database, the manifest will be updated to include the time it was written and this will
+seed the previously returned tick in the clock to the value in the latest manifest:
+
+```fbs
+table ManifestV1 {
+    // ...
+    timestamp: long;
+}
+```
+
+`Db#put_put_with_options` can optionally specify a row-level TTL at insertion time 
+by leveraging newly added field in `WriteOptions`:
+
+```rust
+pub struct WriteOptions {
+    /// the time-to-live (ttl) for this insertion. If this insert overwrites an existing
+    /// database entry, the TTL for the most recent entry will be canonical.
+    ///
+    /// default: the TTL configured in DbOptions when opening a SlateDB session
+    pub ttl: Option<Duration>,
+}
+```
+
+For this RFC we will not introduce the ability to disable filtering out TTL-expired records on
+retrieval. The default implementation will read the metadata for every retrieved row and filter out
+records where `ts + tll < now`. We can, as a future optimization, enable lazy deserialization of the
+metadata so that this doesn't need to happen on each row retrieved if the user requests the ability
+to retrieve TTL'd rows that have not yet been compacted. This can be part of the API design for 
+compaction filters (see https://github.com/slatedb/slatedb/issues/225)
+
+## Implementation
+
+### Data Format
+
+Currently, SlateDB does not have a mechanism for backwards-compatibility of the SST data format.
+
+We will first introduce the following fields to the flatbuffer definitions: 
+- `format_version`: field that specifies the codec version that encoded this SST
+- `meta_features`: an enumeration of metadata fields that are available with each encoded row. This
+  makes it possible in the future to save space per row by disabling metadata features, but that is
+  out of scope for this RFC
+- `creation_timestamp`: the time at which this SST file was written. For this RFC this timestamp is
+  used to schedule follow-up compactions (see [periodic compaction](#periodic-compaction)).
+
+```fbs
+/// the metadata encoded with each row, note that the ordering of this enum is 
+/// sensitive and must be maintained
+enum MetaFeatures: byte {
+    Timestamp,
+    TimeToLive,
+    RowType,
+}
+
+table SsTableInfo {
```

Comment:
I know TWCS is out of scope. Even then, given that we are changing SsTableInfo, could we include the timestamp range included in the Sst? This would avoid another change in the future.

Reply:
Sure thing, can't see the harm in that!